Python/chapter01/1.7 - Rotate Matrix/miguel_1.7_sol.py

```diff
@@ -0,0 +1,174 @@
+"""
+Python version 3.7.0
+1.7 - Rotate Matrix
+Given an image represented by an NxN matrix, where each pixel in the image is 4 bytes, write a method to rotate
+the image by 90 degrees.  Can you do this in place?
+"""
+import unittest
+from typing import List
+
+
+def rotate_matrix(matrix: List[List[int]], N: int) -> List[List[int]]:
+    """
+    Rotate matrix will rotate the given matrix by 90 degrees.
+    Runtime: O(N^2), asymptotic runtime depends on N
+    Space Complexity: O(N^2), creating a new matrix of NxN called 'rotated'
+    :param matrix: an NxN matrix
+    :param N: the size of the matrix (NxN)
+    :return: a newly rotated matrix
+    """
+    rotated = [[0 for i in range(N)] for j in range(N)]
+    for i in range(N):
+        for j in range(N):
+            rotated[j][(N-1)-i] = matrix[i][j]
+    return rotated
+
+
+def rotate_matrix_in_place(matrix: List[List[int]], start_row: int, start_col: int, N: int) -> List[List[int]]:
```

Comment:
you can provide defaults to all parameters except `matrix`:

```python
def rotate_matrix_in_place(
    matrix: List[List[int]],
    start_row: int = 0,
    start_col: int = 0,
    N: Optional[int]) -> List[List[int]]:
  """
  ...
  """
  if N is None:
    N = len(matrix)
```

that way you can invoke both methods exactly the same!